### PR TITLE
Fix configurations for using custom CA.

### PR DESCRIPTION
**Pull request by myidpt**
Please provide a description for what this PR is for.
Fixing two bugs:
1. Setting TrustDomain CA configuration into istio-agent SDS cache.
2. Use the `features.XDSAuth` to control client TLS certificate authentication enablement.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ X ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/bootstrap/server.go [hunk 1]
```diff
@@ -586,9 +586,10 @@ func (s *Server) initSecureDiscoveryService(args *PilotArgs, port string) error
 		return err
 	}
 
+	// TODO(myidpt): Be able to load custom CA certificates to verify clients.
 	cfg := &tls.Config{
 		GetCertificate: s.getIstiodCertificate,
-		ClientAuth:     tls.VerifyClientCertIfGiven,
+		ClientAuth:     tls.NoClientCert,
```
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:592]: I suppose this is fine since we use our own authentication further down the call that Costin added?

Comment (howardjohn) [reply pilot/pkg/bootstrap/server.go:592]: although your description seems wrong that it "has no effect"?

Comment (myidpt) [reply pilot/pkg/bootstrap/server.go:592]: Yes you are right. We defer the authentication to later. Updated.

File: pkg/istio-agent/sds-agent.go [hunk 2]
```diff
@@ -524,6 +524,7 @@ func applyEnvVars() {
 	workloadSdsCacheOptions.SecretTTL = secretTTLEnv
 	workloadSdsCacheOptions.SecretRotationGracePeriodRatio = secretRotationGracePeriodRatioEnv
 	workloadSdsCacheOptions.RotationInterval = secretRotationIntervalEnv
+	workloadSdsCacheOptions.TrustDomain = trustDomainEnv
```
Comment (JimmyCYJ) [inline pkg/istio-agent/sds-agent.go:527]: Just curious, do you know why this does not exist? Was it removed?

**Comment by mandarjog** [comment]
@myidpt tests are failing.
@costinm can you look at this too?

File: pilot/pkg/bootstrap/server.go (hunk 1 above)
Comment (myidpt) [reply pilot/pkg/bootstrap/server.go:591]: Actually VerifyClientCertIfGiven must be present in order for the later authenticator to use the authenticated identity.

File: pkg/istio-agent/sds-agent.go (hunk 2 above)
Comment (myidpt) [reply pkg/istio-agent/sds-agent.go:527]: I think the code was refactored recently.

File: pilot/pkg/bootstrap/server.go (hunk 1 above)
Comment (costinm) [reply pilot/pkg/bootstrap/server.go:591]: I'm not sure this is correct - it will break the authentication using mTLS.

**Comment by costinm** [comment]
Tests are failing because the PR breaks - working as intended.

**Comment by myidpt** [comment]
@costinm why? I added `features.XDSAuth` to control the TLS client authn.

File: pkg/istio-agent/sds-agent.go (hunk 2 above)
Comment (costinm) [reply pkg/istio-agent/sds-agent.go:527]: Isn't trust domain in ProxyConfig ?

**Comment by costinm** [comment]
> @costinm why? I added `features.XDSAuth` to control the TLS client authn.

There is no scenario where Istio will work correctly if this is set to not request certificates.

**Comment by costinm** [comment]
To be clear: the TLS settings control if the TLS ServerHelo includes the bit to request client to send its certificate. 

The client may choose to not do that - and the server will still do the handshake, and not verify the certificate. The authz code can check if a cert was provided ( and if so - it is correct ). 

If the ServerHelo doesn't set this bit - the client will not send a cert, even if it has one.

**Comment by myidpt** [comment]
I understand that what you said, but I don't understand why you think this PR isn't correct.

When the `features.XDSAuth` flag is set, client authentication is required, so the serverHelo will set the bit. Client sends the certificate if it has. The server authenticator in ads works.

When the `features.XDSAuth` flag isn't set, the serverHelo don't set the bit. Client doesn't send the cert at all. The server authenticator is skipped as well.

**Comment by howardjohn** [comment]
even without that flag we still use the authentication for VMs. Why not
just include the client cert but not validate it here?

**Comment by myidpt** [comment]
The "VerifyClientCertIfGiven" is the issue. We don't have the CA cert for authenticating the client cert. It fails in bootstrap/server.go, not even going to ads.

**Comment by howardjohn** [comment]
You can use RequireAnyClientCert which will request it but not validate

**Comment by mandarjog** [comment]
Superseded by https://github.com/istio/istio/pull/24669

**Pull request closed without merging**